### ISSUE TITLE
Bug : SharedMountPoint not working after StorPool plugin support

### ISSUE THREAD

**Issue by dhslove** [apache/cloudstack#6455]
After adding support for StorPool in CloudStack version 4.17, Default Primary Storage of SharedMountPoint Type is no longer created. This Bug is related and caused by #6007 

This is an issue caused by the StoragePoolType of the StorPoolStorageAdaptor.

Currently, the Pool Type of StorPoolStorageAdaptor is set to SharedMountPoint.
See the following code.

StorPoolStorageAdaptor : 
~~~
@StorageAdaptorInfo(storagePoolType=StoragePoolType.SharedMountPoint)
public class StorPoolStorageAdaptor implements StorageAdaptor {
    public static void SP_LOG(String fmt, Object... args) {
    ......
~~~

For this reason, even if it is a Default SharedMountPoint type of storage other than StorPool, LibvirtStorageAdaptor is not selected by the following code, but StorPoolStorageAdaptor is selected and an error occurs.

KVMStoragePoolManager :
~~~
    private StorageAdaptor getStorageAdaptor(StoragePoolType type) {
        // type can be null: LibVirtComputingResource:3238
        if (type == null) {
            return _storageMapper.get("libvirt");
        }
        StorageAdaptor adaptor = _storageMapper.get(type.toString());
        if (adaptor == null) {
            // LibvirtStorageAdaptor is selected by default
            adaptor = _storageMapper.get("libvirt");
        }
        return adaptor;
    }
~~~

In my opinion, it is better to create a separate Pool Type for StorPoolStorageAdaptor like the LINSTOR Plugin.

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
UI, API, Plugin
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0
~~~

##### CONFIGURATION

3 Hosts, GFS2 Clustered File System, Shared Mount Point : /mnt/gfs2

##### OS / ENVIRONMENT

CentOS 8.5, CloudStack 4.17.0

##### STEPS TO REPRODUCE

~~~
1. AddPrimaryStorage 
2. Select SharedMountPoint Protocol
3. Input Mount Directory
4. Select DefaultPrimary Provider
5. Submit 
~~~

##### EXPECTED RESULTS

Create Shared Mount Point Primary Storage

##### ACTUAL RESULTS

SharedMountPoint Primary Storage is recorded in DB, but the pool is not actually created because an appropriate StorageAdaptor is not selected.

**Comment by slavkap** [comment]
@dhslove, unfortunately, a few storage pools are using one Storage adaptor, and the method you mentioned shouldn't rely only on the `StoragePoolType`. I've missed that this could cause any issues, for which I'm sorry. We will work on a fix for 4.17.1.0.
Till then, as a workaround that comes to my mind, you could delete the jars files of the StorPool plug-in for the management/agents. 
```
/usr/share/cloudstack-management/lib/cloud-plugin-storage-volume-storpool-4.17.0.0.jar
/usr/share/cloudstack-agent/lib/cloud-plugin-storage-volume-storpool-4.17.0.0.jar
```
and restart the services
CS doesn't depend on that code, and deleting it shouldn't cause any other issues

**Comment by nvazquez** [comment]
Thanks @slavkap - @dhslove can you please confirm if this workaround solves the issue for you?

**Comment by dhslove** [comment]
@slavkap Thank you for your comment.  

I've already read your README file and solved it as you say. 

For other users, I expect this bug to be resolved in the next patch. 

Thank you for your hard work.

**Comment by dhslove** [comment]
@nvazquez Of course. I would be happy to do so if I can help you with your workaround.

**Comment by rohityadavcloud** [comment]
@dhslove do upgraded env continue to work, for. ex if you had a sharemountpoint based storage prior to the upgrade does it continue to work post upgrade to 4.17.0.0. (I understand the original bug report is adding new sharedmountpoint).

@nvazquez I think if the workaround solves the issue then we should document that part of the release notes (update the docs website) and send an email to both dev and users ML.

**Comment by slavkap** [comment]
@rohityadavcloud, I checked the code, and the method `getStorageAdaptor` is used in almost every case. Before StorPool, if you use SharedMountPoint, it's not added to `_storageMapper` because there isn't a specific adaptor for it, and it uses the libvirt storage adapter. It should get in here:
```
        StorageAdaptor adaptor = _storageMapper.get(type.toString());
        if (adaptor == null) {
            // LibvirtStorageAdaptor is selected by default
            adaptor = _storageMapper.get("libvirt");
        }
```
I'm working to change our storage pool type to StorPool in a way that will not affect current clients. But I think the best solution is to rewrite the `getStorageAdaptor` in cases where different adaptors use `SharedMountPoint`.

**Comment by rohityadavcloud** [comment]
Ping Hi @slavkap are you working on a bugfix PR for this, for 4.17.1.0?

**Comment by slavkap** [comment]
Hi @rohityadavcloud, yes, I'm working on a fix for 4.17.1.0. I need some time for more extensive testing

**Comment by rohityadavcloud** [comment]
Thanks for updating @slavkap

**Comment by mlsorensen** [comment]
Glad to see this is being worked on @slavkap, thanks! I ran into this just today.

Yes, this `@StorageAdaptorInfo(storagePoolType=StoragePoolType.SharedMountPoint)` is used to select which adaptor should be called to implement agent volume requests. As such StorPool probably needs to register their own type instead of trying to be the implementation for SharedMountPoint.

**Comment by nvazquez** [comment]
Hi @mlsorensen yes, unfortunately this error has not been detected earlier. In the meantime, the error can be mitigated by removing the storpool jar from the KVM hosts and the management servers, and restarting the CloudStack services - https://blogs.apache.org/cloudstack/entry/advisory-apache-cloudstack-advisory-on

**Comment by shwstppr** [comment]
@dhslove thanks for testing the PR
Fixed with #6552
